### PR TITLE
Breakage debug flags

### PR DESCRIPTION
**Pull request by muodov**
https://app.asana.com/0/0/1205073793816810/f

This PR adds an MVP infrastructure for breakage debug flags. Currently this is only passed in web extensions (see https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/2138).
In this PR, flags are sent by `fingerprinting-battery` and `fingerprinting-hardware` features, more will come in the follow-up PR.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205073793816810

### CODE REVIEW

File: build-output.eslintrc [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
     "root": true,
-    "plugins": ["promise", "@typescript-eslint"],
+    "plugins": ["promise", "@typescript-eslint", "eslint-plugin-n"],
```
Comment (jonathanKingston) [inline build-output.eslintrc:3]: I still dunno why this was needed hah, has there been anything changed you can see that correlates?

Comment (jonathanKingston) [reply build-output.eslintrc:3]: > error  Definition for rule 'n/no-callback-literal' was not found  n/no-callback-literal

was the error from [this run](https://github.com/duckduckgo/content-scope-scripts/actions/runs/5646734760/job/15295293432?pr=648)

File: inject/chrome.js [hunk 2]
```diff
@@ -13,6 +13,7 @@ const allowedMessages = [
     'getClickToLoadState',
     'getYouTubeVideoDetails',
     'openShareFeedbackPage',
+    'addDebugFlag',
```
Comment (jonathanKingston) [inline inject/chrome.js:16]: Is this still needed, aren't you using the other transport?

File: src/content-feature.js [hunk 3]
```diff
@@ -87,14 +90,32 @@ export default class ContentFeature {
      */
     get messagingContext () {
         if (this.#messagingContext) return this.#messagingContext
+
+        const contextName = import.meta.injectName === 'apple-isolated'
```
Comment (jonathanKingston) [inline src/content-feature.js:94]: We could just land the messaging refactoring + extension support as an isolated PR and discuss the other parts after.

File: src/content-feature.js [hunk 4]
```diff
@@ -222,4 +243,14 @@ export default class ContentFeature {
     // eslint-disable-next-line @typescript-eslint/no-empty-function
     update () {
     }
+
+    /**
+     * Register a string flag that will be added to page breakage reports
+     * @param {string} flag
+     */
+    addDebugFlag (flag = 'fired') {
+        this.debugMessaging?.notify('addDebugFlag', {
```
Comment (jonathanKingston) [inline src/content-feature.js:252]: This method **needs** to memoise and fire only once, to prevent swamping the native messaging with debugging if the site does something like:

```
const pixels = ctx.getImageData(0, 0, canvas.width, canvas.height);
for (let i = 0; i<imageData.data.length; i++) {
  if (navigator.deviceMemory) {
  ...
  }
}
```

**Review by jonathanKingston — CHANGES_REQUESTED** [review]

File: build-output.eslintrc (hunk 1 above)
Comment (muodov) [reply build-output.eslintrc:3]: This test was failing since the first push, I'm not sure what's causing it. I can try to bisect

File: inject/chrome.js (hunk 2 above)
Comment (muodov) [reply inject/chrome.js:16]: This is needed for MV2. We haven't implemented a proper messaging transport for the extension.

Comment (jonathanKingston) [reply inject/chrome.js:16]: Ok thanks for the heads up :)

File: build-output.eslintrc (hunk 1 above)
Comment (jonathanKingston) [reply build-output.eslintrc:3]: And not locally right? So weird. If you get chance it would be worth a bash.

File: unit-test/content-feature.js [hunk 5]
```diff
@@ -49,4 +49,50 @@ describe('Helpers checks', () => {
...
+            feature.addDebugFlag('someflag')
+            expect(spyNotify).toHaveBeenCalledWith(
+                'addDebugFlag',
+                {
+                    flag: 'someFeatureName.someflag'
+                }
+            )
+        })
+
+        it('should not send duplicate flags', () => {
```
Comment (jonathanKingston) [inline unit-test/content-feature.js:78]: Nice thanks for this.

File: src/content-feature.js [hunk 6]
```diff
@@ -222,4 +246,17 @@ export default class ContentFeature {
     // eslint-disable-next-line @typescript-eslint/no-empty-function
     update () {
     }
+
+    /**
+     * Register a string flag that will be added to page breakage reports
+     * @param {string} flag
+     */
+    addDebugFlag (flag = '') {
```
Comment (jonathanKingston) [inline src/content-feature.js:254]: If we're not intending to use this flag code path immediately I would prefer we remove it.

The code instead could set itself to be a noop even (not sure if that's a weird convention and hard to follow) but instead it could check a boolean state instead of a set.

File: src/features/fingerprinting-hardware.js [hunk 7]
```diff
@@ -4,16 +4,25 @@ import { wrapProperty } from '../wrapper-utils'
 export default class FingerprintingHardware extends ContentFeature {
     init () {
         wrapProperty(globalThis.Navigator.prototype, 'keyboard', {
-            // @ts-expect-error - error TS2554: Expected 2 arguments, but got 1.
-            get: () => this.getFeatureAttr('keyboard')
+            get: () => {
+                this.addDebugFlag()
```
[the remark this answers is not in the document]
Comment (muodov) [reply src/features/fingerprinting-hardware.js:8]: I am going to follow up with this change in another PR: https://github.com/duckduckgo/content-scope-scripts/pull/652

File: build-output.eslintrc (hunk 1 above)
Comment (muodov) [reply build-output.eslintrc:3]: I'm getting this error locally even on the current `main` branch. Maybe some of the external packages changed?

**Review by jonathanKingston — APPROVED** [review]

File: build-output.eslintrc (hunk 1 above)
Comment (jonathanKingston) [reply build-output.eslintrc:3]: Not sure (shouldn't do as the packages should be pinned), or the tests were skipped recently. Let's add for now. Thanks for checking.
